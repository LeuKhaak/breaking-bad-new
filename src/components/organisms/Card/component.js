import React from "react";
import * as styles from "./styles.module.scss";
import Loader from "../../../components/atoms/Loader";
import Error from "../../../components/atoms/Error";
import { Link } from "react-router-dom";
import Btn from "../../../components/atoms/Btn";
import arrow from "src/assets/icons/arrow.svg";

function Card({
  personData,
  getError,
  loader,
  quoteData,
  quoteError,
  quoteLoader,
}) {
  return getError ? (
    <Error notice="Character not found!" />
  ) : loader ? (
    <Loader />
  ) : personData ? (
    <div className={styles.card}>
      <div className={styles.container}>
        <Link to="/catalog" className={styles.link}>
          <Btn btnText="Back to catalog" icon={arrow} />
        </Link>
        <div className={styles.person}>
          <div className={styles.image}>
            <img className={styles.img} src={personData.img} alt="#" />
          </div>
          <div className={styles.inf}>
            <span
              className={
                personData.status === "Alive" ? styles.alive : styles.deceased
              }
            >
              {personData.status.toLowerCase()}
            </span>
            <h3 className={styles.name}>{personData.name}</h3>
            <span className={styles.birthday}>
              Birthday:
              <span className={styles.inner}>{personData.birthday}</span>
            </span>
            <span className={styles.niсkName}>
              Niсk-name:
              <span className={styles.inner}>{personData.nickname}</span>
            </span>
            {quoteError ? (
              <Error />
            ) : quoteLoader ? (
              <Loader />
            ) : (
              <span className={styles.quote}>
                <span className={styles.quoteLabel}>Quote:</span>
                <span className={styles.inner}>{quoteData}</span>
              </span>
            )}
          </div>
        </div>
      </div>
    </div>
  ) : (
    <div className={styles.card}>
      <div className={styles.noCharacter}>
        <Error notice="Invalid character name!" />
        <span className={styles.instruction}>
          Go back to the catalog and enter a different name.
        </span>
      </div>
    </div>
  );
}

export default Card;
