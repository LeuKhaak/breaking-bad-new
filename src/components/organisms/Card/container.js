import { useSelector, useDispatch } from "react-redux";
import { React, useEffect } from "react";
import {
  getPersonData,
  clearPersonData,
} from "../../../store/actions/personDetails";
import { getQuoteData, clearQuoteData } from "src/store/actions/quote";
import CardList from "./component";
import { useLocation } from "react-router-dom";

function CardContainer() {
  const dispatch = useDispatch();
  const location = useLocation().pathname.slice(8);

  const personData = useSelector((state) => state.personPage.personData[0]);
  const getError = useSelector((state) => state.personPage.getError);
  const loader = useSelector((state) => state.personPage.loader);

  const quoteData = useSelector((state) => state.quote.randomQuote);
  const quoteError = useSelector((state) => state.quote.quoteError);
  const quoteLoader = useSelector((state) => state.quote.quoteLoader);

  const quote = personData && personData.name ? quoteData : "";

  useEffect(() => {
    dispatch(getPersonData(location));
    return () => {
      dispatch(clearPersonData());
      dispatch(clearQuoteData());
    };
  }, [dispatch, location]);

  useEffect(() => {
    if (personData && personData.name) {
      dispatch(getQuoteData(personData.name));
    }
  }, [dispatch, personData]);

  return (
    <CardList
      quoteData={quote}
      quoteLoader={quoteLoader}
      quoteError={quoteError}
      personData={personData}
      getError={getError}
      loader={loader}
    />
  );
}

export const container = CardContainer;
