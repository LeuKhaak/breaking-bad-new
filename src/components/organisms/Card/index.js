export * from "./component";
export * from "./component";
export { container as default } from "./container";
