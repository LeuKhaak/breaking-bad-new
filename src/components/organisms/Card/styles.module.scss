@import "../../../assets/styles/styles.scss";
.card {
  padding: 2.5rem 0 10rem 0;
  background-color: $bgcCard;
  min-height: calc(100vh - 10rem);
  .container {
    @include content;
    max-width: 74.5rem;
    .link {
      text-decoration: none;
    }
    .person {
      display: flex;
      margin-top: 3rem;
      @include mobile {
        flex-direction: column;
        align-items: center;
      }
      .image {
        width: 36.5rem;
        height: 36.5rem;
        overflow: hidden;
        @include tabletLandscape {
          width: 28rem;
          height: 28rem;
        }
        @include tabletSmall {
          width: 23rem;
          height: 23rem;
        }
        @include tabletSmallSecond {
          width: 18rem;
          height: 18rem;
        }
        @include mobile {
          width: 23rem;
          height: 23rem;
          margin-bottom: 1rem;
        }
        @include mobileSmall {
          width: 18rem;
          height: 18rem;
        }
        .img {
          width: 100%;
        }
      }
    }
    .inf {
      margin-left: 2.75rem;
      max-width: 25rem;
      @include tabletLandscape {
        max-width: 18rem;
      }
      @include tabletSmall {
        max-width: 16rem;
      }
      @include tabletSmallSecond {
        margin-left: 1rem;
      }
      @include mobile {
        max-width: 16rem;
        display: flex;
        flex-direction: column;
        align-items: center;
      }
      .deceased {
        @include status;
        margin-bottom: 2rem;
        background-color: $bgcDeceased;
      }
      .alive {
        @include status;
        background-color: $bgcAlive;
        margin-bottom: 2rem;
        color: $aliveColor;
      }
      .name {
        font-weight: 500;
        font-size: 2rem;
        color: $mainFontColor;
        margin-bottom: 1rem;
      }
      .birthday,
      .niсkName,
      .quote {
        display: block;
        font-size: 1rem;
        color: $greyFontColor;
        .inner {
          color: $mainFontColor;
          padding-left: 0.3rem;
        }
      }
      .birthday {
        margin-bottom: 1rem;
      }
      .niсkName {
        margin-bottom: 2rem;
      }
    }
  }
  .noCharacter {
    text-align: center;
    .instruction {
      display: block;
      font-size: 1rem;
      color: $mainFontColor;
    }
  }
}
