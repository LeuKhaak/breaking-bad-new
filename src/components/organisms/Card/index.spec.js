import React from "react";
import { shallow } from "enzyme";
import Card from "./component";

let component = null;
const setComponent = (props) => shallow(<Card {...props} />);
beforeEach(() => {
  component = setComponent();
});

describe("Card testing:", () => {
  test("is Error shown", () => {
    component = setComponent({
      getError: true,
    });
    expect(component).toMatchSnapshot();
  });

  test("is Loader shown", () => {
    component = setComponent({
      getError: null,
      loader: true,
    });
    expect(component).toMatchSnapshot();
  });

  test("is Card rendered with status Alive", () => {
    component = setComponent({
      getError: null,
      loader: false,
      tiles: true,
      personData: {
        status: "Alive",
        name: "Jeck",
      },
    });
    expect(component).toMatchSnapshot();
  });

  test("is Card rendered with status Deceased", () => {
    component = setComponent({
      getError: null,
      loader: false,
      tiles: true,
      personData: {
        status: "Deceased",
        name: "Ann",
      },
    });
    expect(component).toMatchSnapshot();
  });

  test("is quoteError shown", () => {
    component = setComponent({
      quoteError: true,
    });
    expect(component).toMatchSnapshot();
  });

  test("is quoteLoader shown", () => {
    component = setComponent({
      quoteError: null,
      quoteLoader: true,
    });
    expect(component).toMatchSnapshot();
  });

  test("is quote rendered", () => {
    component = setComponent({
      personData: {
        status: "Alive",
        name: "Jeck",
      },
      quoteError: null,
      quoteLoader: false,
      quoteData: "Test quote",
    });
    expect(component).toMatchSnapshot();
  });
});
