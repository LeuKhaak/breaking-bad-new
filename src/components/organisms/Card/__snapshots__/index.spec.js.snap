// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card testing: is Card rendered with status Alive 1`] = `
<div>
  <div>
    <Link
      to="/catalog"
    >
      <Btn
        btnText="Back to catalog"
        icon="test-file-stub"
      />
    </Link>
    <div>
      <div>
        <img
          alt="#"
        />
      </div>
      <div>
        <span>
          alive
        </span>
        <h3>
          Jeck
        </h3>
        <span>
          Birthday:
          <span />
        </span>
        <span>
          Niсk-name:
          <span />
        </span>
        <span>
          <span>
            Quote:
          </span>
          <span />
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card testing: is Card rendered with status Deceased 1`] = `
<div>
  <div>
    <Link
      to="/catalog"
    >
      <Btn
        btnText="Back to catalog"
        icon="test-file-stub"
      />
    </Link>
    <div>
      <div>
        <img
          alt="#"
        />
      </div>
      <div>
        <span>
          deceased
        </span>
        <h3>
          Ann
        </h3>
        <span>
          Birthday:
          <span />
        </span>
        <span>
          Niсk-name:
          <span />
        </span>
        <span>
          <span>
            Quote:
          </span>
          <span />
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card testing: is Error shown 1`] = `
<Error
  notice="Character not found!"
/>
`;

exports[`Card testing: is Loader shown 1`] = `<Loader />`;

exports[`Card testing: is quote rendered 1`] = `
<div>
  <div>
    <Link
      to="/catalog"
    >
      <Btn
        btnText="Back to catalog"
        icon="test-file-stub"
      />
    </Link>
    <div>
      <div>
        <img
          alt="#"
        />
      </div>
      <div>
        <span>
          alive
        </span>
        <h3>
          Jeck
        </h3>
        <span>
          Birthday:
          <span />
        </span>
        <span>
          Niсk-name:
          <span />
        </span>
        <span>
          <span>
            Quote:
          </span>
          <span>
            Test quote
          </span>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Card testing: is quoteError shown 1`] = `
<div>
  <div>
    <Error
      notice="Invalid character name!"
    />
    <span>
      Go back to the catalog and enter a different name.
    </span>
  </div>
</div>
`;

exports[`Card testing: is quoteLoader shown 1`] = `
<div>
  <div>
    <Error
      notice="Invalid character name!"
    />
    <span>
      Go back to the catalog and enter a different name.
    </span>
  </div>
</div>
`;
